compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirFunctionParameterChecker.kt

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.declaration
+
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.declarations.isInline
+import org.jetbrains.kotlin.fir.typeContext
+import org.jetbrains.kotlin.fir.types.arrayElementType
+import org.jetbrains.kotlin.fir.types.coneType
+import org.jetbrains.kotlin.fir.types.isUnsignedTypeOrNullableUnsignedType
+import org.jetbrains.kotlin.types.AbstractTypeChecker
+
+object FirFunctionParameterChecker : FirFunctionChecker() {
+    override fun check(declaration: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        checkVarargParameters(declaration, context, reporter)
+    }
+
+    private fun checkVarargParameters(function: FirFunction<*>, context: CheckerContext, reporter: DiagnosticReporter) {
+        val varargParamters = function.valueParameters.filter { it.isVararg }
+        if (varargParamters.size > 1) {
+            for (parameter in varargParamters) {
+                reporter.reportOn(parameter.source ?: continue, FirErrors.MULTIPLE_VARARG_PARAMETERS, context)
+            }
+        }
+
+        val nullableNothingType = context.session.builtinTypes.nullableNothingType.coneType
+        for (varargParameter in varargParamters) {
+            val varargParameterType = varargParameter.returnTypeRef.coneType.arrayElementType() ?: continue
+            if (AbstractTypeChecker.isSubtypeOf(context.session.typeContext, varargParameterType, nullableNothingType) ||
+                (varargParameterType.isInline(context.session) && !varargParameterType.isUnsignedTypeOrNullableUnsignedType)
+            // Here we skip checking if the type is not primitive because primitive types are not inline. That is any primitive values are
```

Comment:
This comment is a little confusing to read. I don't understand why we need to add a comment about checking primitive types, it's just more confusing to me.

Reply:
This is a note about the difference of this implementation compared with FE1.0. Let me mention that in the comment.